File: sdk/villow-0.1.0/villow-0.1.0/src/villow/signing.py
from __future__ import annotations

import hashlib
import hmac
import secrets
import time
from dataclasses import dataclass


KEY_ID_HEADER = "x-worklane-key-id"
SIGNATURE_HEADER = "x-worklane-signature"
TIMESTAMP_HEADER = "x-worklane-timestamp"
NONCE_HEADER = "x-worklane-nonce"
PUBLISHER_ID_HEADER = "x-worklane-publisher-id"
AGENT_ID_HEADER = "x-worklane-agent-id"
IDEMPOTENCY_KEY_HEADER = "x-worklane-idempotency-key"


@dataclass(frozen=True)
class SigningPrincipal:
    publisher_id: str
    agent_id: str
    key_id: str
    nonce: str
    timestamp: int
    idempotency_key: str


def request_body_hash(body: bytes) -> str:
    return hashlib.sha256(body).hexdigest()


def canonical_request(
    method: str,
    path: str,
    body: bytes,
    *,
    timestamp: int,
    nonce: str,
    key_id: str,
    publisher_id: str,
    agent_id: str,
    idempotency_key: str,
) -> str:
    return "\n".join(
        [
            method.upper(),
            path,
            str(timestamp),
            nonce,
            key_id,
            publisher_id,
            agent_id,
            idempotency_key,
            request_body_hash(body),
        ]
    )


def sign_request(
    method: str,
    path: str,
    body: bytes,
    *,
    key_id: str,
    secret: str,
    publisher_id: str,
    agent_id: str,
    idempotency_key: str,
    timestamp: int | None = None,
    nonce: str | None = None,
) -> dict[str, str]:
    timestamp = int(timestamp or time.time())
    nonce = nonce or secrets.token_hex(16)
    canonical = canonical_request(
        method,
        path,
        body,
        timestamp=timestamp,
        nonce=nonce,
        key_id=key_id,
        publisher_id=publisher_id,
        agent_id=agent_id,
        idempotency_key=idempotency_key,
    )
    digest = hmac.new(secret.encode("utf-8"), canonical.encode("utf-8"), hashlib.sha256).hexdigest()
    return {
        KEY_ID_HEADER: key_id,
        SIGNATURE_HEADER: f"v1:{digest}",
        TIMESTAMP_HEADER: str(timestamp),
        NONCE_HEADER: nonce,
        PUBLISHER_ID_HEADER: publisher_id,
        AGENT_ID_HEADER: agent_id,
        IDEMPOTENCY_KEY_HEADER: idempotency_key,
    }


def verify_request_signature(
    method: str,
    path: str,
    body: bytes,
    headers: dict[str, str],
    *,
    secret: str,
    now: int | None = None,
    max_skew_seconds: int = 300,
) -> SigningPrincipal:
    normalized = {key.lower(): value for key, value in headers.items()}
    try:
        key_id = normalized[KEY_ID_HEADER]
        signature = normalized[SIGNATURE_HEADER]
        timestamp = int(normalized[TIMESTAMP_HEADER])
        nonce = normalized[NONCE_HEADER]
        publisher_id = normalized[PUBLISHER_ID_HEADER]
        agent_id = normalized.get(AGENT_ID_HEADER, "")
        idempotency_key = normalized.get(IDEMPOTENCY_KEY_HEADER, "")
    except (KeyError, ValueError) as exc:
        raise ValueError("missing_or_invalid_signature_headers") from exc

    now = int(now or time.time())
    if abs(now - timestamp) > max_skew_seconds:
        raise ValueError("stale_signature_timestamp")

    canonical = canonical_request(
        method,
        path,
        body,
        timestamp=timestamp,
        nonce=nonce,
        key_id=key_id,
        publisher_id=publisher_id,
        agent_id=agent_id,
        idempotency_key=idempotency_key,
    )
    expected = "v1:" + hmac.new(secret.encode("utf-8"), canonical.encode("utf-8"), hashlib.sha256).hexdigest()
    if not hmac.compare_digest(signature, expected):
        raise ValueError("invalid_signature")
    return SigningPrincipal(publisher_id, agent_id, key_id, nonce, timestamp, idempotency_key)

File: sdk/villow-sdk-0.1.0/package/src/signing.ts
import { createHash, createHmac, timingSafeEqual } from 'node:crypto';

export const KEY_ID_HEADER = 'x-worklane-key-id';
export const SIGNATURE_HEADER = 'x-worklane-signature';
export const TIMESTAMP_HEADER = 'x-worklane-timestamp';
export const NONCE_HEADER = 'x-worklane-nonce';
export const PUBLISHER_ID_HEADER = 'x-worklane-publisher-id';
export const AGENT_ID_HEADER = 'x-worklane-agent-id';
export const IDEMPOTENCY_KEY_HEADER = 'x-worklane-idempotency-key';

export interface SigningPrincipal {
  publisherId: string;
  agentId: string;
  keyId: string;
  nonce: string;
  timestamp: number;
  idempotencyKey: string;
}

export interface SignRequestInput {
  method: string;
  path: string;
  body: Buffer | string | Uint8Array;
  keyId: string;
  secret: string;
  publisherId: string;
  agentId: string;
  idempotencyKey: string;
  timestamp?: number;
  nonce?: string;
}

export interface VerifyRequestInput {
  method: string;
  path: string;
  body: Buffer | string | Uint8Array;
  headers: Record<string, string | string[] | undefined>;
  secret: string;
  now?: number;
  maxSkewSeconds?: number;
}

export function requestBodyHash(body: Buffer | string | Uint8Array): string {
  return createHash('sha256').update(toBuffer(body)).digest('hex');
}

export function canonicalRequest(input: Omit<SignRequestInput, 'secret'> & { timestamp: number; nonce: string }): string {
  return [
    input.method.toUpperCase(),
    input.path,
    String(input.timestamp),
    input.nonce,
    input.keyId,
    input.publisherId,
    input.agentId,
    input.idempotencyKey,
    requestBodyHash(input.body),
  ].join('\n');
}

export function signRequest(input: SignRequestInput): Record<string, string> {
  const timestamp = input.timestamp ?? Math.floor(Date.now() / 1000);
  const nonce = input.nonce ?? createHash('sha256').update(`${timestamp}:${Math.random()}`).digest('hex').slice(0, 32);
  const canonical = canonicalRequest({ ...input, timestamp, nonce });
  const digest = createHmac('sha256', input.secret).update(canonical).digest('hex');
  return {
    [KEY_ID_HEADER]: input.keyId,
    [SIGNATURE_HEADER]: `v1:${digest}`,
    [TIMESTAMP_HEADER]: String(timestamp),
    [NONCE_HEADER]: nonce,
    [PUBLISHER_ID_HEADER]: input.publisherId,
    [AGENT_ID_HEADER]: input.agentId,
    [IDEMPOTENCY_KEY_HEADER]: input.idempotencyKey,
  };
}

export function verifyRequestSignature(input: VerifyRequestInput): SigningPrincipal {
  const headers = normalizeHeaders(input.headers);
  const keyId = requireHeader(headers, KEY_ID_HEADER);
  const signature = requireHeader(headers, SIGNATURE_HEADER);
  const timestamp = Number.parseInt(requireHeader(headers, TIMESTAMP_HEADER), 10);
  const nonce = requireHeader(headers, NONCE_HEADER);
  const publisherId = requireHeader(headers, PUBLISHER_ID_HEADER);
  const agentId = headers[AGENT_ID_HEADER] ?? '';
  const idempotencyKey = headers[IDEMPOTENCY_KEY_HEADER] ?? '';
  if (!Number.isInteger(timestamp)) {
    throw new Error('missing_or_invalid_signature_headers');
  }

  const now = input.now ?? Math.floor(Date.now() / 1000);
  const maxSkewSeconds = input.maxSkewSeconds ?? 300;
  if (Math.abs(now - timestamp) > maxSkewSeconds) {
    throw new Error('stale_signature');
  }

  const canonical = canonicalRequest({
    method: input.method,
    path: input.path,
    body: input.body,
    keyId,
    publisherId,
    agentId,
    idempotencyKey,
    timestamp,
    nonce,
  });
  const expected = `v1:${createHmac('sha256', input.secret).update(canonical).digest('hex')}`;
  if (!safeEqual(signature, expected)) {
    throw new Error('invalid_signature');
  }
  return { publisherId, agentId, keyId, nonce, timestamp, idempotencyKey };
}

function normalizeHeaders(headers: Record<string, string | string[] | undefined>): Record<string, string> {
  const normalized: Record<string, string> = {};
  for (const [key, value] of Object.entries(headers)) {
    if (typeof value === 'string') {
      normalized[key.toLowerCase()] = value;
    } else if (Array.isArray(value) && value.length > 0) {
      normalized[key.toLowerCase()] = value[0];
    }
  }
  return normalized;
}

function requireHeader(headers: Record<string, string>, name: string): string {
  const value = headers[name];
  if (!value) {
    throw new Error('missing_or_invalid_signature_headers');
  }
  return value;
}

function safeEqual(actual: string, expected: string): boolean {
  const actualBuffer = Buffer.from(actual);
  const expectedBuffer = Buffer.from(expected);
  return actualBuffer.length === expectedBuffer.length && timingSafeEqual(actualBuffer, expectedBuffer);
}

function toBuffer(body: Buffer | string | Uint8Array): Buffer {
  return Buffer.isBuffer(body) ? body : Buffer.from(body);
}
